# 예외처리

### 예외와 오류의 차이

**오류(Error)**

- **시스템 레벨**의 심각한 문제점
- 컴파일 시 문법적인 오류와 런타임 시 널포인트 참조와 같은 오류이다.
- 프로세스에 심각한 문제를 야기시켜 프로세스를 종료시킬 수 있다.
- **복구할 수 없는 심각한 오류**이다.

**예외(Exception)**

- **프로그램 레벨**의 일반적인 문제점
- 컴퓨터 시스템의 동작 도중 예기치 않았던 이상 상태가 발생하여 수행 중인 프로그램이 영향을 받는 것이다.
    - null 처리
    - 공백 검사
    - 예외 처리
- **발생하더라도 수습할 수 있는 비교적 덜 심각한 오류**이다.  
<br>
### 예외 처리

자바에서는 예외와 에러도 객체로 처리한다.

**예외가 주로 발생하는 원인**

- 사용자의 잘못된 데이터 입력
- 잘못된 입력
- 개발자가 로직을 잘못 작성한 경우
- 하드웨어, 네트워크 오작동
- 시스템 과부하

### Throwable 클래스

- 예외처리를 할 수 있는 최상위 클래스이다.
- Exception과 Error는 Throwable의 상속을 받는다.  
<br>
### 오류(Error)

![java-error-exception-003](https://github.com/user-attachments/assets/f49e922a-1272-4153-afe7-8a39264fc977)

- 시스템 레벨에서 발생하여, 개발자가 어떻게 조치할 수 없는 수준

### 예외(Exception)

![java-error-exception-002](https://github.com/user-attachments/assets/de08075c-2af0-48fa-a00a-6638e2723d41)

- 개발자가 구현한 로직에서 발생하여 개발자가 다른 방식으로 처리가능한 것들로 JVM은 정상동작한다.  

<br>

**Exception의 2가지 종류**

1. Checked Exception
    - 예외처리가 필수이며, 처리하지 않으면 컴파일 되지 않는다.
    - JVM 외부와 통신(네트워크, 파일시스템 등)할때 주로 쓰인다.
    - RuntimeExcaption 이외에 있는 모든 예외
    - IOException, SQLException 등
2. Unchecked Exception
    - 컴파일 때 체크되지 않고, Runtime에 발생하는 Excpetion을 말한다.
    - RuntimeException 하위의 모든 예외
    - NulPointerException, IndexOutofBoundException 등  

<br>

**대표적인 Exception 클래스**

| 클래스명 | 설명 |
| --- | --- |
| NullPointerException |  Null 레퍼런스를 참조할때 발생, 뭔가 동작시킬 때 발생 |
| IndexOutOfBoundsException | 배열과 유사한 자료구조(문자열, 배열, 자료구조)에서 범위를 벗어난 인덱스 번호 사용으로 발생 |
| FormatException | 문자열, 숫자, 날짜 변환 시 잘못된 데이터(ex. "123A" -> 123 으로 변환 시)로 발생하며, 보통 사용자의 입력, 외부 데이터 로딩, 결과 데이터의 변환 처리에서 자주 발생 |
| ArthmeticException | 정수를 0으로 나눌때 발생 |
| ClassCastException | 변환할 수 없는 타입으로 객체를 변환할 때 발생 |
| IllegalArgumentException | 잘못된 인자 전달 시 발생 |
| IOException | 입출력 동작 실패 또는 인터럽트 시 발생 |
| IllegalStateException | 객체의 상태가 매소드 호출에는 부적절한 경우에 발생 |
| ConcurrentModificationException | 금지된 곳에서 객체를 동시에 수정하는것이 감지될 경우 발생 |
| UnsupportedOperationException | 객체가 메소드를 지원하지 않는 경우 발생 |

<br>

**주요 메소드**

`printStackTrace()` 

- 발생한 Exception의 출처를 메모리 상에서 추적하면서 결과를 알려준다.
- 발생한 위치를 정확히 출력해주므로 가장 많이 사용되며, void를 반환한다.

`getMessage()` 

- 한줄로 요약된 메시지를 String 타입으로 반환한다.

`getStackTrace()`

- jdk1.4부터 지원, printStackTrace()를 보완, StackTraceElement[] 이라는 문자열 배열로 변경해서 출력하고 저장한다.  

<br>

### **예외 처리 방식**

Java에서 모든 예외가 발생하면 (XXX)Exception 객체를 생성한다. 이때 예외 처리 방법은 크게 2가지로 나뉜다.

1. 직접 try~catch를 이용해서 예외에 대한 최종적인 책임을 지고 처리한다.
2. throws Exception을 이용해서 발생한 예외의 책임을 호출하는 쪽이 책임지도록 한다.
    - 주로 호출하는 쪽에 예외를 보고할 때 사용한다.

다른 메소드의 일부분으로 동작하는 경우에는 던지는 것을 추천한다.

1. **try-catch**
    - 발생한 에러를 던지면 catch가 잡아서 해결한다.
    - try문 안에는 에러가 발생하는 위험한 로직이 들어가고, catch문 안에는 예외 발생시 수행할 로직이 들어간다.
    - try중이라도 예외가 발생한 다음의 코드들은 실행되지 않으며, catch 구문으로 넘어간다.
    - catch구문은 else if 처럼 여러개 쓸 수 있다.
    - finally는 마지막에 실행하고 싶은 로직이 들어가며, 대표적으로 .close()가 있다.
    
    예시)
    
    ```java
    try {
          String userInput =  "12.3";
          
          int number = Integer.parseInt(userInput);
          System.out.println(number);
          
        } catch(NumberFormatException e) {
          System.out.println("정수만 입력하세요.");
        }
    ```
    
    - 위의 코드에서는 NumberFormatException에러가 날것으로 예상되는 실행문을 try문안에 넣고 괄호 안에는 에러의 종류를 적어준다.
    - 해당 실행문에서 에러가 발생하면 catch문 안의 내용으로 예외 처리가 된다.

1. **throw**
    - 직접 예외 객체를 만들어서 던질 때 사용한다.
    - 자바가 자동으로 던지지 않는 예외들은 throw문으로 직접 던져야 한다.
    
    예시) 
    
    ```java
    try {
          
          int score = 100; // 0 ~ 100
          char grade;
    
          if(score >= 90) {
            grade = 'A';
          } else if(score >= 80) {
            grade = 'B';
          } else if(score >= 70) {
            grade = 'C';
          } else if(score >= 60) {
            grade = 'D';
          } else 
            grade = 'F';
          
          System.out.println("점수 : " + score + "점, 학점 : " + grade + "학점");
          
        } catch (Exception e) {
          
          System.out.println("잘못된 점수입니다.");
          
        }
    ```
    
    - 만약 score 값이 음수나 100보다 큰 숫자가 들어온다면 우리(개발자)는 잘못된 값으로 판단한다.
    - 그러나 자바에게는? int score에 -1이 들어오는 것은 지극히 정상적인 결과이다.
    - 그러므로 이를 throw문을 이용해서 처리해주자.
    
    ```java
    if(score < 0 || score > 100) {
            throw new RuntimeException();
    }
    ```
    
    - 코드 중간에 if문을 이용해서 에러가 나는 score의 조건(음수이거나 100점보다 큰 값)을 설정해주고 이 에러를 throw 키워드를 통해 예외 객체를 생성해서 던진다.
        - RuntimeException 추천.
    
    ```java
    catch (Exception e) {
          System.out.println("잘못된 점수입니다.");
    }
    ```
    
    - 위에서 던진 에러 객체는 catch가 받는다.
        
        Exception e(매개변수) = new RuntimeException(throw를 통해 던진 인자) 합치면 객체 생성 구문이 된다.
        

1. **throws**
    - 메소드에서 발생한 예외를 메소드를 호출한 곳으로 던질 때 사용한다.
    - 2개 이상의 예외를 던질 수 있다.
    - 메소드를 호출한 곳에서는 예외 처리 코드가 추가되어야 한다.
    
    예시)
    
    ```java
    public void division(int a, int b) throws ArithmeticException {
        System.out.println(a + "+" + b + "=" + (a / b));
      }
    ```
    
    - 해당 메소드에서 throws를 이용해서 예상되는 에러 클래스를 명시한다. 에러가 발생하면 catch로 바로 에러를 던진다.
    
    ```java
    calculator.division(a, b);
    ```
    
    - 메소드가 호출될 때 a / b는 5/0임으로 에러가 뜬다.
    
    ```java
    try {      
          Calculator calculator = new Calculator();
          calculator.addition(a, b);
          calculator.subtraction(a, b);
          calculator.multiplication(a, b);
          calculator.division(a, b);      
        } catch (ArithmeticException e) {
          System.out.println("산술 연산 예외가 발생했습니다.");
        }
    ```
    
    - try catch로 예외처리를 해준다.
    

문제1) 오류와 예외의 차이점을 설명해보세요.

문제2) 예외 처리 방법에는 무엇이 있나요?

---

**참고자료**

[https://gyoogle.dev/blog/computer-language/Java/Error & Exception.html](https://gyoogle.dev/blog/computer-language/Java/Error%20&%20Exception.html)